http/etag.ts

```diff
@@ -0,0 +1,196 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+
+/** Provides functions for dealing with and matching ETags, including
+ * {@linkcode calculate} to calculate an etag for a given entity,
+ * {@linkcode ifMatch} for validating if an ETag matches against a `If-Match`
+ * header and {@linkcode ifNoneMatch} for validating an Etag against an
+ * `If-None-Match` header.
+ *
+ * See further information on the `ETag` header on
+ * [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag).
+ *
+ * @module
+ */
+
+import { type DigestAlgorithm } from "../crypto/crypto.ts";
+import { toHashString } from "../crypto/to_hash_string.ts";
+import { createHash } from "../crypto/_util.ts";
+
+/** Just the part of `Deno.FileInfo` that is required to calculate an `ETag`,
+ * so partial or user generated file information can be passed. */
+export interface FileInfo {
+  mtime: Date | null;
+  size: number;
+}
+
+type Entity = string | Uint8Array | FileInfo;
+
+const DEFAULT_ALGORITHM = "FNV32A";
+const ENV_PERM_STATUS =
+  Deno.permissions.querySync?.({ name: "env", variable: "DENO_DEPLOYMENT_ID" })
+    .state ?? "granted"; // for deno deploy
+const DENO_DEPLOYMENT_ID = ENV_PERM_STATUS === "granted"
+  ? Deno.env.get("DENO_DEPLOYMENT_ID")
+  : undefined;
+const HASHED_DENO_DEPLOYMENT_ID = DENO_DEPLOYMENT_ID
+  ? createHash("FNV32A", DENO_DEPLOYMENT_ID).then((hash) => toHashString(hash))
+  : undefined;
+
+export interface ETagOptions {
+  /** A digest algorithm to use to calculate the etag. Defaults to
+   * `"FNV32A"`. */
+  algorithm?: DigestAlgorithm;
+
+  /** Override the default behavior of calculating the `ETag`, either forcing
+   * a tag to be labelled weak or not. */
+  weak?: boolean;
+}
+
+function isFileInfo(value: unknown): value is FileInfo {
+  return Boolean(
+    value && typeof value === "object" && "mtime" in value && "size" in value,
+  );
+}
+
+async function calcEntity(
+  entity: string | Uint8Array,
+  { algorithm }: ETagOptions,
+) {
+  // a short circuit for zero length entities
+  if (!entity.length && !algorithm) {
+    return "811c9dc5";
+  }
+
+  return toHashString(await createHash(algorithm ?? DEFAULT_ALGORITHM, entity));
+}
+
+async function calcFileInfo(fileInfo: FileInfo, { algorithm }: ETagOptions) {
+  if (fileInfo.mtime) {
+    return toHashString(
+      await createHash(
+        algorithm ?? DEFAULT_ALGORITHM,
+        `${fileInfo.mtime.toJSON()}${fileInfo.size}`,
+      ),
+    );
+  }
+  return HASHED_DENO_DEPLOYMENT_ID;
+}
+
+/** Calculate an ETag for an entity. When the entity is a specific set of data
+ * it will be fingerprinted as a "strong" tag, otherwise if it is just file
+ * information, it will be calculated as a weak tag.
+ *
+ * ```ts
+ * import { calculate } from "https://deno.land/std@$STD_VERSION/http/etag.ts";
+ * import { assert } from "https://deno.land/std@$STD_VERSION/_util/asserts.ts"
```

Comment:
I don't think we should recommend that users use std's internal assert.

```suggestion
 * import { assert } from "https://deno.land/std@$STD_VERSION/testing/asserts.ts"
```